sdks/python/apache_beam/runners/sdf_utils.py

```diff
@@ -244,3 +248,68 @@ def get_estimator_state(self):
         return None
 
     return _NoOpWatermarkEstimator()
+
+
+class DeduplicationByUniqueId(ptransform.PTransform):
+  """A Transform that perform deduplication based on the unique id.
+
+  The input should be a pair of (key, value), where key stands for the unique_id
+  of the value. The transform outputs the value with orignal windowing strategy.
+  By default, the deduplication will be performed with Session windows. The
+  default size of Session is one day.
+  """
+  def __init__(self, session_size=24 * 60 * 60):
+    self._session_size = session_size
+
+  class OutputValueWithOriginalWindowing(core.ParDo):
+    class _DropUniqueId(core.DoFn):
+      def process(self, element):
+        _, windowed_value = element
+        yield windowed_value
+
+    def __init__(self, original_windowing):
+      self.windowing = original_windowing
+      super(DeduplicationByUniqueId.OutputValueWithOriginalWindowing,
+            self).__init__(self._DropUniqueId())
+
+    def get_windowing(self, unused_inputs):
+      return self.windowing
+
+  class _DeduplicationCombineFn(core.CombineFn):
+    def create_accumulator(self):
+      return []
+
+    def add_input(self, accumulator, element):
+      if len(accumulator) == 0:
+        accumulator.append(element)
+      return accumulator
+
+    def merge_accumulators(self, accumulators):
+      for accumulator in accumulators:
+        if len(accumulator) > 0:
+          return accumulator
+
+    def extract_output(self, accumulator):
+      return accumulator[0]
+
+  class KeepValueWindowingInfo(core.DoFn):
+    def process(
+        self,
+        kv,
+        ts=core.DoFn.TimestampParam,
+        window=core.DoFn.WindowParam,
+        paneinfo=core.DoFn.PaneInfoParam):
+      id, value = kv
+      yield (id, WindowedValue(value, ts, [window], paneinfo))
```

Comment:
The deduplication key should also contain the window since we don't want to remove output that may have been assigned to multiple windows. Most SDFs will produce output in the global window.